sqlalchemy_bigquery/_struct.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Mapping, Tuple
+
+import packaging.version
+import sqlalchemy.sql.default_comparator
+import sqlalchemy.sql.sqltypes
+import sqlalchemy.types
+
+from . import base
+
+sqlalchemy_1_4_or_more = packaging.version.parse(
+    sqlalchemy.__version__
+) >= packaging.version.parse("1.4")
+
+if sqlalchemy_1_4_or_more:
+    import sqlalchemy.sql.coercions
+    import sqlalchemy.sql.roles
+
+# We have to delay getting the type compiler, because of circular imports. :(
+type_compiler = None
+
+
+class STRUCT(sqlalchemy.sql.sqltypes.Indexable, sqlalchemy.types.UserDefinedType):
+    def __init__(
+        self,
+        *fields: Tuple[str, sqlalchemy.types.TypeEngine],
+        **kwfields: Mapping[str, sqlalchemy.types.TypeEngine],
+    ):
+        self.__fields = tuple(
```

Comment:
Are we relying on one or the other superclasses to do something with these fields? Let's comment if so.

Also, should we be `__init__`ing one or both of our super classes?

Reply:
> Are we relying on one or the other superclasses to do something with these fields? Let's comment if so.

No.

> Also, should we be `__init__`ing one or both of our super classes?

No.

See: https://docs.sqlalchemy.org/en/14/core/custom_types.html#creating-new-types

and also: [ARRAY](https://github.com/sqlalchemy/sqlalchemy/blob/06f031e55f40403996e3580bdc5a22661b6303f4/lib/sqlalchemy/sql/sqltypes.py#L2661), which subclasses  `Indexable` and other things but doesn't call superclass `__init__`s.

We could call `super().__init__`, which would feel hygienic :), but would be a noop and might make the reader wonder what it's doing.  I would be OK either way.

Reply:
`__fields` is completely specific to `STRUCT` (this the leading `__`).